### ISSUE TITLE
Upstreaming and pull request strategy for Spark on Kubernetes

### ISSUE THREAD

**Issue by erikerlandson** [apache-spark-on-k8s/spark#441]
We'll be wanting a plan for presenting this project in PRs that are as digestible as possible by the Apache Spark reviewers upstream.  We can use this issue to formulate a strategy

**Comment by erikerlandson** [comment]
The RSS seems like one good candidate for a PR, to be presented "after" the kube-spark core.  If so, then possibly RSS-related config might require temporary disabling for initial core PR.

**Comment by erikerlandson** [comment]
The hadoop repo seems like another.

**Comment by erikerlandson** [comment]
Shuffle service / D.A. ?

**Comment by foxish** [comment]
The candidate set of different and separable components, and a potential order in which we could upstream them is: 

1. Test infrastructure (minikube on jenkins)
2. Scheduler Backend + Submission Client
3. Dynamic allocation
4. Resource staging server
5. Kerberos support

(2) is the most complex of the lot. It mostly makes sense together, but we could break it down further if we absolutely must.

**Comment by foxish** [comment]
Following up on the discussion today in the meeting, the questions we need to answer are:

**Test infrastructure**
- Discovery - What's the right place to run our integration tests? AMPLab Jenkins? Externally? - @ssuchter 
- Current resource requirements & running tests in parallel. - @mccheah  
- Can we run tests on an external k8s cluster (is this okay from the upstream perspective?) @ssuchter/@foxish to follow up. 

**Tag 2.2 branch**
- Merge fixes
- Cut a bug-fix release and announce. (@kimoonkim)

**Scheduler Backend + Submission Client**
* Steps
  - Submission orchestrator
  - Submission steps
  - Scheduler Backend
  - (@mcheah to assess if this makes sense for now, or if we more of a split)

cc @felixcheung

**Comment by mccheah** [comment]
It's also good to document the changes that we want to include in upstream that we haven't merged into branch-2.2-kubernetes yet, including:

- Resilience to SPARK-21642 - #482, PR #483 
- Move shuffle directory mounting for executors to a separate class - PR #454 
- Unit tests for the scheduler backend - PR #459 

Any other ones I'm missing?

**Comment by foxish** [comment]
@mccheah all of those have merged I think.

For the 2.3 code-freeze, this seems to me like a reasonable (maybe slightly ambitious) target - 
I expect we can speed up after the first PR.

- [x] **PR 1**: Basic scheduler backend PR (https://github.com/apache-spark-on-k8s/spark/pull/498, https://github.com/apache/spark/pull/19468)
    - [x] Unit testing for executorpodfactory (https://github.com/apache-spark-on-k8s/spark/pull/491)
- [x] **PR 2**: Basic submission client PR - covers spark-submit changes and Client.scala + the minimum set of submission steps and orchestrator (https://github.com/apache/spark/pull/19717, https://github.com/apache-spark-on-k8s/spark/pull/545)
  - [x] Improve Unit testing for SparkSubmit changes
  - [x] Unit tests for Client.scala
  - [x] Reference Dockerfiles for driver and executor.
- [x] **PR 2.5**: Docker images
  - [ ] ASF procedure for pre-built images
- [x] **PR 3**: Init-container setup + make-distribution setup
  - [x] Improve Unit testing (TBD)
- [ ] **PR 4**: Dynamic allocation PR
  - [ ] Improve Unit testing (TBD)
- [x] **PR 6**: Documentation
- [x] **PR ?** Parallel effort on Integration Tests
  - [x] Discovery - RiseLAB CI
  - [ ] Will the ASF allow us to run on external GKE/GCE cluster? 

Thoughts? If we all agree, I'll post this on the JIRA.

**Comment by erikerlandson** [comment]
Per 9/20 sig meeting, move docker images up to after PR2.  Also, annotate that Pr3 (submission steps) will be partly component of PR2.

In general, I'd expect that specific submission steps would be added along with whatever components they're associated with.  The basic submission step architecture would land with PR2.

**Comment by felixcheung** [comment]
re: https://github.com/apache-spark-on-k8s/spark/issues/441#issuecomment-327655209
how about SparkR support? PR # 507
